backends/ref/ceed-ref-operator.c

```diff
@@ -1109,8 +1109,212 @@ static int CeedOperatorLinearAssembleQFunctionAtPointsUpdate_Ref(CeedOperator op
+  CeedCallBackend(CeedOperatorGetNumElements(op, &num_elem));
+  CeedCallBackend(CeedOperatorGetQFunction(op, &qf));
+  CeedCallBackend(CeedOperatorGetFields(op, &num_input_fields, &op_input_fields, &num_output_fields, &op_output_fields));
+  CeedCallBackend(CeedQFunctionGetFields(qf, NULL, &qf_input_fields, NULL, &qf_output_fields));
+
+  // Setup
+  CeedCallBackend(CeedOperatorSetupAtPoints_Ref(op));
+
+  // Ceed
+  {
+    Ceed ceed_parent;
+
+    CeedCallBackend(CeedOperatorGetCeed(op, &ceed));
+    CeedCallBackend(CeedGetParent(ceed, &ceed_parent));
+    if (ceed_parent) ceed = ceed_parent;
+  }
+
+  // Point coordinates
+  CeedCallBackend(CeedOperatorAtPointsGetPoints(op, &rstr_points, &point_coords));
+
+  // Input and output vectors
+  {
+    CeedSize input_size, output_size;
+
+    CeedCallBackend(CeedOperatorGetActiveVectorLengths(op, &input_size, &output_size));
+    CeedCallBackend(CeedVectorCreate(ceed, input_size, &in_vec));
+    CeedCallBackend(CeedVectorCreate(ceed, output_size, &out_vec));
+    CeedCallBackend(CeedVectorSetValue(out_vec, 0.0));
+  }
+
+  // Input Evecs and Restriction
+  CeedCallBackend(CeedOperatorSetupInputs_Ref(num_input_fields, qf_input_fields, op_input_fields, NULL, true, e_data, impl, request));
+
+  // Check if active field is at points
+  for (CeedInt i = 0; i < num_input_fields; i++) {
+    CeedRestrictionType rstr_type;
+    CeedVector          vec;
+    CeedElemRestriction elem_rstr;
+
+    CeedCallBackend(CeedOperatorFieldGetVector(op_input_fields[i], &vec));
+    // Skip non-active input
+    if (vec != CEED_VECTOR_ACTIVE) continue;
+
+    // Get active restriction type
+    CeedCallBackend(CeedOperatorFieldGetElemRestriction(op_input_fields[i], &elem_rstr));
+    CeedCallBackend(CeedElemRestrictionGetType(elem_rstr, &rstr_type));
+    CeedCallBackend(CeedElemRestrictionGetNumComponents(elem_rstr, &num_comp_active));
+    is_active_at_points = rstr_type == CEED_RESTRICTION_POINTS;
+    if (!is_active_at_points) CeedCallBackend(CeedElemRestrictionGetElementSize(elem_rstr, &elem_size_active));
+  }
+
+  // Loop through elements
+  for (CeedInt e = 0; e < num_elem; e++) {
+    CeedInt num_points, e_vec_size = 0;
+
+    // Setup points for element
+    CeedCallBackend(CeedElemRestrictionApplyAtPointsInElement(rstr_points, e, CEED_NOTRANSPOSE, point_coords, impl->point_coords_elem, request));
+    CeedCallBackend(CeedElemRestrictionGetNumPointsInElement(rstr_points, e, &num_points));
+
+    // Input basis apply for non-active bases
+    CeedCallBackend(CeedOperatorInputBasisAtPoints_Ref(e, num_points_offset, num_points, qf_input_fields, op_input_fields, num_input_fields, in_vec,
+                                                       impl->point_coords_elem, true, e_data, impl, request));
+
+    // Loop over points on element
+    e_vec_size = (is_active_at_points ? num_points : elem_size_active) * num_comp_active;
+    for (CeedInt s = 0; s < e_vec_size; s++) {
```

Comment:
Just to clarify, for our typical Jacobian operator at points, this is a loop over the nodes of each element that computes the diagonal contributions for that node from the points within that element, right?

I want to make sure we aren't doing something with like `O(num_points^2)`  complexity.

Reply:
For our typical approach, we assemble the QF by repeated application of D [1, 0, 0...] and then compute the matrix triple product B D B^T and extract the element operator matrix diagonal. (we might be doing something clever there so we don't compute the full matrix triple product)

Here we are instead repeatedly applying B D B^T [1, 0, 0...] to build the element operator matrix diagonal.

This PR should be more expensive than our typical approach when we have consistent quadrature points in each element but less expensive than building B for every element

Reply:
Gotcha, sounds good to me.